### PR TITLE
Consume leading zeroes when parsing a number in ConstantExpressionTemplate

### LINKED ISSUE #54380: Incorrect handling of leading zeroes during INSERT
**How to reproduce**
```sql
CREATE TABLE EMPLOYEE
(
    `empId` INTEGER PRIMARY KEY,
    `d` DOUBLE
)
ENGINE = MergeTree
ORDER BY empId;

INSERT INTO EMPLOYEE VALUES (0001, 1.456), (0005, 45.98), (0008, 4342.766), (0017, 345.87), (0021, 43.78), (0051, 0.781);

SELECT * FROM EMPLOYEE;
```
Results into
```
┌─empId─┬────────d─┐
│     1 │    1.456 │
│     5 │    45.98 │
│     8 │ 4342.766 │
│    15 │   345.87 │
│    17 │    43.78 │
│    41 │    0.781 │
└───────┴──────────┘
```
**Expected behavior**
```
┌─empId─┬────────d─┐
│     1 │    1.456 │
│     5 │    45.98 │
│     8 │ 4342.766 │
│    17 │   345.87 │
│    21 │    43.78 │
│    51 │    0.781 │
└───────┴──────────┘
```

Reproduces on the latest master:
https://fiddle.clickhouse.com/8d27f2fe-31a8-4af5-a54b-c0d1608886db

### PR DESCRIPTION
**Pull request by jh0x**
Fixes #54380

In most cases the octals are being dealt with properly (i.e. leading zeroes are consumed and numbers are being parsed as non-octals - in `ExpressionElementParsers`). This addresses a use-case in `ConstantExpressionTemplate`.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not interpret numbers with leading zeroes as octals

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 14b23e3df1122b4ea5504cf2e0dc874d1573f406 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55575/14b23e3df1122b4ea5504cf2e0dc874d1573f406/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6500764347">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55575/14b23e3df1122b4ea5504cf2e0dc874d1573f406/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@jh0x see `test_format_schema_on_server`

**Comment by alexey-milovidov** [comment]
@jh0x it needs additional attention.

**Comment by jh0x** [comment]
> @jh0x it needs additional attention.

I updated the branch and see:
- `Integration tests (asan) [1/4] — fail: 1, passed: 644` -> seems unrealated - stacktraces in pqxx?
- `Integration tests (tsan) [6/6] — fail: 1, passed: 495` -> seems unrelated - stacktraces in kafka?
- `Performance Comparison [2/4] — 2 errors, 2 too long, 2 unstable` -> is there anything in particular that I should investigate?

**Comment by alexey-milovidov** [comment]
Yes, the Postgres library had a memory leak - we have fixed it recently.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
> [The query no. 0 of test 'insert_values_with_expressions' is taking too long to run. Keep the run time below 2 seconds"](https://s3.amazonaws.com/clickhouse-test-reports/55575/da8f210caa34743f9a7e58670ccf14130bb13247/performance_comparison_[2_4]/report.html#backward-incompatible-queries.insert_values_with_expressions.0)

File: src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp [hunk 1]
```diff
@@ -597,6 +597,10 @@ bool ConstantExpressionTemplate::parseLiteralAndAssertType(
             if (negative || *istr.position() == '+')
                 ++istr.position();
 
+            /// Consume leading zeroes - we don't want any funny octal business
```
Comment (alexey-milovidov) [inline src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp:600]: The loop by bytes is slow. It has to be optimized.

Comment (alexey-milovidov) [reply src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp:600]: The slowness could be in the case when we speculatively try to parse a number, but it is a string starting with a lot of zeros.

**Comment by alexey-milovidov** [comment]
> is there anything in particular that I should investigate?

Yes, the performance degraded, so we have to bring the performance back.

**Comment by alexey-milovidov** [comment]
Something wrong with the performance test `insert_values_with_expressions`

File: src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp [hunk 2]
```diff
@@ -603,18 +603,23 @@ bool ConstantExpressionTemplate::parseLiteralAndAssertType(
             memcpy(buf, istr.position(), bytes_to_copy);
             buf[bytes_to_copy] = 0;
 
-            char * pos_double = buf;
+            /// Consume leading zeroes - we don't want any funny octal business
+            auto* non_zero_buf = buf;
+            while (*non_zero_buf == '0')
```
Comment (alexey-milovidov) [inline src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp:608]: It could be slow for long sequences of zeros.
Check `find_first_not_symbols`.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by tavplubix** [comment]
@jh0x, @alexey-milovidov, this PR breaks the parsing of numbers with ConstantExpressionTemplate and perf test `insert_values_with_expressions`: https://s3.amazonaws.com/clickhouse-test-reports/58931/21e089b02754ea0f4043d3048a25825f2d17382b/performance_comparison_[1_4]/report.html#run-errors

The test added in this PR does not test anything because it does not disable `input_format_values_interpret_expressions`, so it falls back to `evaluateConstantExpression`. 

The problem: if the number is `0`, then the new code skips this zero, so there's nothing to parse anymore, and it fails to parse a number.

Sorry, I have to revert this change. @jh0x, please resubmit it with a proper test.

**Comment by jh0x** [comment]
@tavplubix Thank you, I will have a look.

**Comment by alexey-milovidov** [comment]
@jh0x let's continue.

**Pull request merged**